### PR TITLE
Definitions for properties-reader.

### PR DESCRIPTION
**Pull request by Goldsmith42**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/xsd-schema-validator/tsconfig.json [hunk 1]
```diff
@@ -19,4 +19,4 @@
         "index.d.ts",
         "xsd-schema-validator-tests.ts"
     ]
-}
```
Comment (andy-ms) [inline types/xsd-schema-validator/tsconfig.json]: Don't change unrelated files.

File: types/duplexer2/tsconfig.json [hunk 2]
```diff
@@ -5,6 +5,7 @@
     ],
     "compilerOptions": {
         "module": "commonjs",
+        "target": "es6",
```
Comment (andy-ms) [inline types/duplexer2/tsconfig.json]: Why is this change needed?

File: types/properties-reader/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// Definitions by: Zlatko Andonovski <https://github.com/Goldsmith42/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare namespace PropertiesReader {
+    type Value = string|number|boolean;
+
+    interface Reader {
+        get(propertyName: string): Value|null;
+        getRaw(propertyName: string): string|null;
+        path(): Object;
```
Comment (andy-ms) [inline types/properties-reader/index.d.ts]: Might as well just return `any` here.

File: types/properties-reader/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        append(path: string): Reader;
+        read(properties: string): Reader;
+        set(propertyName: string, value: Value): Reader;
+        length: number;
+        each(iterator: (key: string, value: Value) => void, scope?: Object): Reader;
```
Comment (andy-ms) [inline types/properties-reader/index.d.ts]: What is `scope`?

File: types/xsd-schema-validator/tsconfig.json (hunk 1 above)
Comment (Goldsmith42) [inline types/xsd-schema-validator/tsconfig.json]: Ugh, looks like some unrelated files from a previous commit got in. I'm going to exclude those when I resubmit (since the CI build failed).

File: types/properties-reader/index.d.ts (hunk 4 above)
Comment (Goldsmith42) [inline types/properties-reader/index.d.ts]: It's a binding variable to use in place of `this`. I pulled the name from the module's source code.

Comment (andy-ms) [inline types/properties-reader/index.d.ts]: If you want support for a this-parameter you could add an overload:
```ts
each(iterator: (key: string, value: Value) => void): Reader;
each(iterator: (this: T, key: string, value: Value) => void, scope?: T): Reader;
```
Be sure to test.

File: types/properties-reader/index.d.ts (hunk 3 above)
Comment (Goldsmith42) [inline types/properties-reader/index.d.ts]: I did some testing with this, it does offer some extra safety if you're trying to do something silly like cast the result to `Function`.

File: types/properties-reader/index.d.ts (hunk 4 above)
Comment (Goldsmith42) [inline types/properties-reader/index.d.ts]: The `this` parameter does not go into the iterator function. Adding an overload for just the `scope` argument causes lint to fail because the signatures can be combined.

Comment (andy-ms) [inline types/properties-reader/index.d.ts]: So where does `this` go?

Comment (Goldsmith42) [inline types/properties-reader/index.d.ts]: It's the context to which the iterator function is bound. What `each` does is that it calls `iterator` for each property. If you use `this` inside iterator, it'll refer to the object provided with `scope`, otherwise it'll be the reader object itself.

Comment (andy-ms) [inline types/properties-reader/index.d.ts]: So, the value of `scope` is the value of `this` inside the iterator function -- that's what the second overload there is trying to say. And it sounds like the first overload should use `this: Reader`. Don't know why tslint is giving you problems there; you could add a `// tslint:disable-next-line unified-signatures`.

**Comment by andy-ms** [comment]
@Goldsmith42 Why did you close this?

File: types/properties-reader/index.d.ts (hunk 4 above)
Comment (Goldsmith42) [inline types/properties-reader/index.d.ts]: My bad, I had a bad understanding of how the fake `this` parameter worked. I will use an overload here.

**Comment by Goldsmith42** [comment]
Easiest way I had of dealing with the extra commits was to reset my fork. This automatically closed the pull request.

**Comment by andy-ms** [comment]
Ah OK. For that reason, it's better to avoid changing the `master` branch, and use a custom branch instead.
